src/main/scala/org/kframework/minikore/interfaces/MiniKoreInterface.scala

```diff
@@ -0,0 +1,303 @@
+package org.kframework.minikore.interfaces
+
+object TreeInterface {
+
+  import PatternInterface._
+
+  sealed trait AST
+
+
+  sealed trait Node extends AST with Product {
+    def args: Seq[Pattern]
+
+    def build(children: Seq[Pattern]): Pattern
+  }
+
+  object Node {
+    def unapply(arg: Node): Option[Seq[Pattern]] = Some(arg.args)
+  }
+
+
+  sealed trait Leaf[C] extends AST with Product {
+    def contents: C
+
+    def build(contents: C): Pattern
+  }
+
+  object Leaf {
+    def unapply[C](arg: Leaf[C]): Option[C] = Some(arg.contents)
+  }
+
+
+  sealed trait Leaf2[CC1, CC2] extends Leaf[Product2[CC1, CC2]] with Product2[CC1, CC2] {
+    override def contents: Product2[CC1, CC2] = (_1, _2)
+
+    def apply(_1: CC1, _2: CC2): Pattern
```

Comment:
The last commit renames the apply methods in the interface to build. I guess this is a better solution since it gives us the leeway to introduce apply methods in the future. Also note that the lines of code from 1->2 is not a good metric, since 2 has many lines of documentation that 1 didn't.